File: backend/app/models/job.py
"""Job posting document."""
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

from app.models.base import MongoModel, PyObjectId
from app.models.enums import DEFAULT_PIPELINE, JobStatus, JobType, Workplace


class Salary(BaseModel):
    min: int | None = None
    max: int | None = None
    currency: str = "INR"
    period: str = "month"  # month | year | hour | stipend


class JobStats(BaseModel):
    views: int = 0
    applicants: int = 0


class Job(MongoModel):
    company_id: PyObjectId
    posted_by: PyObjectId
    title: str
    slug: str
    type: JobType = JobType.INTERNSHIP
    workplace: Workplace = Workplace.REMOTE
    location: str | None = None
    description: str
    responsibilities: list[str] = Field(default_factory=list)
    requirements: list[str] = Field(default_factory=list)
    skills: list[str] = Field(default_factory=list)
    experience_level: str = "entry"
    salary: Salary = Field(default_factory=Salary)
    openings: int = 1
    deadline: datetime | None = None
    status: JobStatus = JobStatus.DRAFT
    pipeline_stages: list[str] = Field(default_factory=lambda: list(DEFAULT_PIPELINE))
    stats: JobStats = Field(default_factory=JobStats)

File: backend/src/models/job.js
"use strict";

/** Job posting document. Mirrors app/models/job.py. */

const mongoose = require("mongoose");
const { Schema } = mongoose;
const { baseSchemaOptions, subSchemaOptions } = require("./plugins");
const { JobType, Workplace, JobStatus, DEFAULT_PIPELINE, values } = require("./enums");

const salarySchema = new Schema(
  {
    min: { type: Number, default: null },
    max: { type: Number, default: null },
    currency: { type: String, default: "INR" },
    period: { type: String, default: "month" }, // month | year | hour | stipend
  },
  subSchemaOptions
);

const statsSchema = new Schema(
  {
    views: { type: Number, default: 0 },
    applicants: { type: Number, default: 0 },
  },
  subSchemaOptions
);

const jobSchema = new Schema(
  {
    company_id: { type: String, required: true },
    posted_by: { type: String, required: true },
    title: { type: String, required: true },
    slug: { type: String, required: true, unique: true },
    type: { type: String, enum: values(JobType), default: JobType.INTERNSHIP },
    workplace: { type: String, enum: values(Workplace), default: Workplace.REMOTE },
    location: { type: String, default: null },
    description: { type: String, required: true },
    responsibilities: { type: [String], default: [] },
    requirements: { type: [String], default: [] },
    skills: { type: [String], default: [] },
    experience_level: { type: String, default: "entry" },
    salary: { type: salarySchema, default: () => ({}) },
    openings: { type: Number, default: 1 },
    deadline: { type: Date, default: null },
    status: { type: String, enum: values(JobStatus), default: JobStatus.DRAFT },
    pipeline_stages: { type: [String], default: () => [...DEFAULT_PIPELINE] },
    stats: { type: statsSchema, default: () => ({}) },
  },
  baseSchemaOptions()
);

jobSchema.index({ status: 1, type: 1, workplace: 1 });
jobSchema.index({ company_id: 1, status: 1 });
jobSchema.index({ skills: 1 });
jobSchema.index({ created_at: -1 });
jobSchema.index({ title: "text", description: "text", skills: "text" });

const Job = mongoose.model("Job", jobSchema, "jobs");

module.exports = { Job };
